### DISCUSSION TITLE
public API: too many requests

### DISCUSSION THREAD

**Discussion by thomaskamalakis** [Budibase/budibase#5679] [Help] [answered]
Hi

  I am getting a weird behavior when using the public API to search for the rows of the table. Here is the python code we use:

```
API_KEY = 'yadayadaya' # We get this value from the "View API key" of an admin account
TABLE_ID = 'ta_whaetever' # This is the table ID
URL = 'https://example.com/api/public/v1/tables/' + TABLE_ID + '/rows/search'
APP_ID = 'app_whatever'

payload = {"query": {"equal": {"email": "thomas@example.com"}}}
headers = {
    "Accept": "application/json",
    "Content-Type": "application/json",
    "x-budibase-api-key": API_KEY,
    'x-budibase-app-id' : APP_ID
}

response = requests.post(URL, json = payload, headers = headers)
print(response.text)
```
  The table with id TABLE_ID is public. When this code is executed, we get a {"message":"Too many requests, please try again later."}

  Here is a part of the docker-compose logs
```
bbapps                | INFO  [1651554818557] (28 on 362e153dd391): request completed
bbapps                |     res: {
bbapps                |       "statusCode": 429,
bbapps                |       "headers": {
bbapps                |         "vary": "Accept-Encoding",
bbapps                |         "set-cookie": [
bbapps                |           "budibase:currentapp=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhcHBJZCI6ImFwcF9kNWI4NzYzOWEwM2Q0NTAwYTM1NWU4MDdkNTdkNDdjYiIsImlhdCI6MTY1MTU1NDgxOH0.I_bti7PEMJsB9FvURv0PmW9hq9sTKyxbHLCWiYc5jIk; path=/; expires=Tue, 19 Jan 2038 03:14:07 GMT"
bbapps                |         ],
bbapps                |         "x-ratelimit-limit": "10",
bbapps                |         "x-ratelimit-remaining": "0",
bbapps                |         "x-ratelimit-reset": "0",
bbapps                |         "content-type": "application/json; charset=utf-8",
bbapps                |         "retry-after": "1",
bbapps                |         "content-length": "56"
bbapps                |       }
bbapps                |     }
bbapps                |     responseTime: 345
bbapps                |     req: {
bbapps                |       "id": 141,
bbapps                |       "method": "POST",
bbapps                |       "url": "/api/public/v1/tables/ta_...../rows/search",
bbapps                |       "headers": {
bbapps                |         "connection": "upgrade",
bbapps                |         "x-real-ip": "my ip",
bbapps                |         "x-forwarded-for": "my pc ip, my web server ip",
bbapps                |         "host": "app-service:4002",
bbapps                |         "content-length": "47",
bbapps                |         "x-forwarded-proto": "https",
bbapps                |         "user-agent": "python-requests/2.22.0",
bbapps                |         "accept-encoding": "gzip, deflate",
bbapps                |         "accept": "application/json",
bbapps                |         "content-type": "application/json",
bbapps                |         "x-budibase-api-key": "my key",
bbapps                |         "x-budibase-app-id": "my app id"
bbapps                |       },
bbapps                |       "remoteAddress": "::ffff:172.18.0.4",
bbapps                |       "remotePort": 53992
bbapps                |     }
```

  Now the thing is that if I change my API_KEY to something gibberish the code works(!) and returns data. However the logs are different in the x-ratelimit-* part:

```
bbapps                |     res: {
bbapps                |       "statusCode": 200,
bbapps                |       "headers": {
...
bbapps                |         "x-ratelimit-limit": "10",
bbapps                |         "x-ratelimit-remaining": "9",
bbapps                |         "x-ratelimit-reset": "1651555309",
bbapps                |         "content-type": "application/json; charset=utf-8",
bbapps                |         "content-length": "1225"
```

What is happening here? My guess is that there is some kind of rate limiting in the bbproxy but I am not sure how to deal with this. I can also confirm that if I create a query inside a REST data source of budibase I get the same behavior.

  Thanks for your help

  Thomas

**Comment by mike12345567** [comment, marked as answer]
Hi @thomaskamalakis - if you are using Budibase Cloud there is no way to change the rate limiting behaviour of Budibase right now.

If you have self hosted Budibase, then you can change the environment variable `API_REQ_LIMIT_PER_SEC` on the `budibase/apps` service to whatever limit you desire. There is also some limits set within the bbproxy service, these should be higher but if you still run into issues they are a little bit more difficult to adjust - this is something we will need to look at, but you can get around it by exposing the apps service directly, rather than contacting it via the proxy, if you add: 
```
ports:
  - "4002:4002"
```
to your `docker-compose.yaml` for the apps service, then you will be able to communicate directly with it on this port and you shouldn't experience any rate limiting issues. I'll make a note that we need to add an environment variable to the proxy image which allows setting the number of requests per second to a higher limit within it as well.

**Reply by vassil-velichkov** [reply]
Same issue with my self-hosted Budibase - trying to create 110K user accounts via API calls from a PHP script, but I hit the bbproxy rate limit every now and then. Editing the nginx.conf file within the docker container, increasing all rate limits 10x, and invoking "/etc/init.d/nginx reload" didn't help.
Would it be sufficient If I just add
ports:
 - "4002:4002"
to the docker-compose.yaml file or I need to change other configurations as well in order to bypass the bbproxy service?

**Reply by shogunpurple** [reply]
@vassil-velichkov did you restart `nginx`? You will need to for those settings to take effect.

**Reply by vassil-velichkov** [reply]
Restarting nginx inside the container restarts the container :-(. I've checked the .js code, it seems the bbproxy nginx.conf file is generated from .hbs template file and the limits are hard-coded.

**Reply by vassil-velichkov** [reply]
usleep() for 500ms in the PHP script alleviates the issue, but it'll take 12 hours to create those 110K users (registered refugees). I guess there are other params related to couchDB/Errlang that might need some tuning. The hardware is just fine, far from loaded...

**Reply by shogunpurple** [reply]
Hi @vassil-velichkov,

So running self hosted, here are the steps you can take to use a completely custom NGINX configuration:

**Update docker-compose.yaml to use a bind mount for your NGINX config**

Replace `/absolute/path/on/server` with wherever you want to store your NGINX configuration.
```
  proxy-service:
    restart: unless-stopped
    ports:
      - "${MAIN_PORT}:10000"
    container_name: bbproxy
    image: budibase/proxy
    # Add the below
    volumes:
      - /absolute/path/on/server:/etc/nginx/nginx.conf
    depends_on:
      - minio-service
      - worker-service
      - app-service
      - couchdb-service
```

**Create the custom NGINX config**
Create an NGINX configuration at the path on your server, by taking the one that is currently in your `proxy-service` container. Here's a one liner that will do it in one command (ensure you are in the directory containing your `docker-compose.yaml`:

`docker-compose exec proxy-service cat /etc/nginx/nginx.conf > nginx.conf`

**Remove the NGINX rate limit**
Remove the following lines:
- `limit_req_zone $binary_remote_addr zone=ratelimit:10m rate=20r/s;`
- `limit_req zone=ratelimit burst=20 nodelay;`

**Restart the bb-proxy container**
`docker-compose up -d`

This should remove the NGINX rate limits altogether.

Please let us know via support@budibase.com if this does not work for you, and we can support you further. It sounds like this app you are building is going to help a lot of people, and we will help you in any way we can to get it up and running.

Thanks!

**Comment by thomaskamalakis** [comment]
Dear @mike12345567 

  Thank you for your feedback. We are on self-hosted budibase (1.0.108). Bashing to the bbapps container, there was no ```API_REQ_LIMIT_PER_SEC``` when I ran the ```env``` command. I did issue a 

```
export API_REQ_LIMIT_PER_SEC=1000
```

  and the variable now appears in ```env``` but nothing has changed. Stop and starting the budi hosting also did not change the behavior. 

I created a new admin user and used his API key instead to change the rows of a table with BASIC read/write access and did hit the request limit after creating ~30 rows, now it also always returns:

```
{"message":"Too many requests, please try again later."}
```
  Is there some sort of user "jail" in place here with the user not being released?

  If I use a dummy key and change the table to PUBLIC the API rows are created fine.

  Thanks!

  T.

**Reply by mike12345567** [reply]
Hi @thomaskamalakis - you need to set the environment variable in your `docker-compose.yaml` as part of the `environment` property, e.g.

```
app-service:
  ...
  environment:
    API_REQ_LIMIT_PER_SEC: 1000
    ...
```

Once you've add this to your docker-compose and apply the changes with `docker-compose up -d` it should then be using that property.

**Reply by ClarkNiu** [reply]
Hi @mike12345567 - I wonder if there is an environment property in Kubernetes scripts?

**Reply by shogunpurple** [reply]
@ClarkNiu it is not in our current kubernetes setup, I've added it here to this PR:

https://github.com/Budibase/budibase/pull/6115

Once merged and released, you will be able to do:

```
  apps:
    port: 4002
    replicaCount: 1
    logLevel: info
    publicApiRateLimitPerSecond: 100000
```

Thank you for the suggestion!

**Reply by ClarkNiu** [reply]
@shogunpurple Thank you so much!

**Reply by ClarkNiu** [reply]
@shogunpurple I have another question. If there is a property that control the maximum number of connections in database connection pool?

**Comment by thomaskamalakis** [comment]
Dear @mike12345567 

  Thank you it seems to be working now!

  T.
